scilpy/tracking/fibertube.py

```diff
@@ -0,0 +1,559 @@
+        sample = np.dot(rotation_matrix, np.append(sample, np.float32(1.)))[:3]
+
+        # Translation
+        sample += center
+
+        bottom_center = center - axis * half_length
+        top_center = center + axis * half_length
+        if (point_in_cylinder(bottom_center, top_center, radius, sample)):
+            samples.append(sample)
+    return samples
+
+
+@njit
+def point_in_cylinder(pt1, pt2, r, q):
+    vec = np.subtract(pt2, pt1)
+    cond_1 = np.dot(q - pt1, vec) >= 0
+    cond_2 = np.dot(q - pt2, vec) <= 0
+    cond_3 = (np.linalg.norm(np.cross(q - pt1, vec)) /
+              np.linalg.norm(vec)) <= r
+    return cond_1 and cond_2 and cond_3
+
+
+@njit
+def sphere_cylinder_intersection(sph_p, sph_r: float, cyl_p1, cyl_p2,
+                                 cyl_r: float, sample_count: int,
+                                 random_generator: np.random.Generator):
+    """
+    Estimates the volume of intersection between a cylinder and a sphere by
+    sampling the cylinder. Most efficient when the cylinder is smaller than
+    the sphere.
+
+    Parameters
+    ----------
+    sph_p: ndarray
+        Center coordinate of the sphere.
+    sph_r: float
+        Radius of the sphere.
+    cyl_p1: ndarray
+        First point of the cylinder's center segment.
+    cyl_p2: ndarray
+        Second point of the cylinder's center segment.
+    cyl_r: float
+        Radius of the cylinder.
+    sample_count: int
+        Amount of samples to use for the estimation.
+
+    Returns
+    -------
+    inter_volume: float
+        Approximate volume of the sphere-cylinder intersection.
+    is_estimated: boolean
+        Indicates whether or not the resulting volume has been estimated.
+        If true, inter_volume has a probability of error due to sample count.
+    """
+    cyl_axis = cyl_p2 - cyl_p1
+    cyl_length = np.linalg.norm(cyl_axis)
+    cyl_center = (cyl_p1 + cyl_p2) / 2
+
+    # If cylinder is completely inside the sphere.
+    if (np.linalg.norm(sph_p - cyl_p1) + cyl_r <= sph_r and
+            np.linalg.norm(sph_p - cyl_p2) + cyl_r <= sph_r):
+        cyl_volume = np.pi * (cyl_r ** 2) * cyl_length
+        return cyl_volume, False
+
+    # If cylinder is completely outside the sphere.
+    _, vector, _, _ = dist_point_segment(cyl_p1, cyl_p2, sph_p)
+    if np.linalg.norm(vector) >= sph_r + cyl_r:
+        return 0, False
+
+    # High probability of intersection.
+    samples = sample_cylinder(cyl_center, cyl_axis, cyl_r, cyl_length,
+                              sample_count, random_generator)
+
+    inter_samples = 0
+    for sample in samples:
+        if np.linalg.norm(sph_p - sample) < sph_r:
+            inter_samples += 1
+
+    # Proportion of cylinder samples common to both shapes * cylinder volume.
+    cyl_volume = np.pi * (cyl_r ** 2) * cyl_length
+    inter_volume = (inter_samples / sample_count) * cyl_volume
+
+    return inter_volume, True
+
+
+@njit
+def create_perpendicular(v):
```

Comment:
add docstring. what happens when v = [0, 0, 0] ?

Reply:
Added handling of v=[0, 0, 0] and docstring